src/github.com/matrix-org/dendrite/encryptoapi/routing/keys.go

```diff
@@ -0,0 +1,537 @@
+	}
+
+	/*
+		federation consideration: when user id is in federation, a query is needed to ask fed for keys
+		domain --------+ fed (keys)
+		domain +--tout-- timer
+	*/
+	// todo: Add federation processing at specific userID.
+	if false /*federation judgement*/ {
+		tout := claimRq.Timeout
+		stimuCh := make(chan int)
+		go func() {
+			time.Sleep(time.Duration(tout) * 1000 * 1000)
+			close(stimuCh)
+		}()
+		select {
+		case <-stimuCh:
+			claimRp.Failures = make(map[string]interface{})
+			// todo: key in this map is restricted to username at the end, yet a mocked one.
+			claimRp.Failures["@alice:localhost"] = "ran out of offered time"
+		case <-make(chan interface{}):
+			// todo : here goes federation chan , still a mocked one
+		}
+	}
+
+	content := claimRq.ClaimDetail
+	for uid, detail := range content {
+		for deviceID, al := range detail {
+			var alTyp int
+			if strings.Contains(al, "signed") {
+				alTyp = ONETIMEKEYOBJECT
+			} else {
+				alTyp = ONETIMEKEYSTRING
+			}
+			key, err := pickOne(req.Context(), *encryptionDB, uid, deviceID, al)
+			if err != nil {
+				claimRp.Failures[uid] = fmt.Sprintf("%s:%s", "fail to get keys for device ", deviceID)
+			}
+			claimRp.ClaimBody[uid] = make(map[string]map[string]interface{})
+			keyPreMap := claimRp.ClaimBody[uid]
+			keymap := keyPreMap[deviceID]
+			if keymap == nil {
+				keymap = make(map[string]interface{})
+			}
+			switch alTyp {
+			case ONETIMEKEYSTRING:
+				keymap[fmt.Sprintf("%s:%s", al, key.KeyID)] = key.Key
+			case ONETIMEKEYOBJECT:
+				sig := make(map[string]map[string]string)
+				sig[uid] = make(map[string]string)
+				sig[uid][fmt.Sprintf("%s:%s", "ed25519", deviceID)] = key.Signature
+				keymap[fmt.Sprintf("%s:%s", al, key.KeyID)] = types.KeyObject{Key: key.Key, Signature: sig}
+			}
+			claimRp.ClaimBody[uid][deviceID] = keymap
+		}
+	}
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: claimRp,
+	}
+}
+
+// todo: check through interface for duplicate and what type of request should it be
+// whether device or one time or both of them
+func checkUpload(req *types.UploadEncryptSpecific, typ int) bool {
+	if typ == BODYDEVICEKEY {
+		devicekey := req.DeviceKeys
+		if devicekey.UserID == "" {
+			return false
+		}
+	}
+	if typ == BODYONETIMEKEY {
+		if req.OneTimeKey.KeyString == nil || req.OneTimeKey.KeyObject == nil {
+			return false
+		}
+	}
+	return true
+}
+
+// QueryOneTimeKeys todo: complete this field through claim type
+func QueryOneTimeKeys(
+	ctx context.Context,
+	typ int,
+	userID, deviceID string,
+	encryptionDB *storage.Database,
+) interface{} {
+	if typ == TYPESUM {
+		res, _ := encryptionDB.SelectOneTimeKeyCount(ctx, deviceID, userID)
+		return res
+	}
+	return nil
+}
+
+// ClearUnused when web client sign out, a clean should be processed, cause all keys would never been used from then on.
+// todo: complete this function and invoke through sign out extension or some scenarios else those matter
+func ClearUnused() {}
```

Comment:
As you are not using this anywhere I would suggest removing this (and only leaving the comment)

Reply:
I'll comment it